File: finding-seams/python/src/speaker_feedback/speaker_feedback.py
from __future__ import annotations

from typing import Literal, TypedDict

from speaker_feedback.feedback_sensor import FeedbackSensor


class EvaluateSpeakerResult(TypedDict):
    status: Literal["ALERT", "BOOK_AGAIN", "OK"]
    averageScore: float | None


class SpeakerFeedbackService:
    def __init__(self) -> None:
        self._sensor = FeedbackSensor()
        self._alert_threshold = 2.2

    def evaluate_speaker(self) -> EvaluateSpeakerResult:
        """
        Collect scores from the unreliable sensor, retrying on failures,
        and classify the speaker.

        Returns an object like:
        { status: 'ALERT' | 'BOOK_AGAIN' | 'OK', averageScore: number | null }
        """
        scores = self._sensor.read_scores()

        if not scores or len(scores) == 0:
            return {
                "status": "ALERT",
                "averageScore": None,
            }

        valid_scores = [
            score
            for score in scores
            if isinstance(score, (int, float)) and 1 <= int(score) <= 5
        ]

        if len(valid_scores) == 0:
            return {
                "status": "ALERT",
                "averageScore": None,
            }

        total = sum(int(s) for s in valid_scores)
        average = total / len(valid_scores)

        if average < self._alert_threshold:
            return {
                "status": "ALERT",
                "averageScore": average,
            }

        return {
            "status": "OK",
            "averageScore": average,
        }

File: finding-seams/typescript/src/speaker-feedback/speaker-feedback.ts
import { FeedbackSensor } from './feedback-sensor.js'

export type EvaluateSpeakerResult = {
  status: 'ALERT' | 'BOOK_AGAIN' | 'OK'
  averageScore: number | null
}

export class SpeakerFeedbackService {
  #sensor: FeedbackSensor
  #alertThreshold = 2.2

  constructor() {
    this.#sensor = new FeedbackSensor()
  }

  /**
   * Collect scores from the unreliable sensor, retrying on failures,
   * and classify the speaker.
   *
   * Returns an object like:
   * { status: 'ALERT' | 'BOOK_AGAIN' | 'OK', averageScore: number | null }
   */
  evaluateSpeaker(): EvaluateSpeakerResult {
    const scores = this.#sensor.readScores()

    if (!scores || scores.length === 0) {
      return {
        status: 'ALERT',
        averageScore: null,
      }
    }

    const validScores = scores.filter(
      score => typeof score === 'number' && score >= 1 && score <= 5,
    )

    if (validScores.length === 0) {
      return {
        status: 'ALERT',
        averageScore: null,
      }
    }

    const sum = validScores.reduce((acc, value) => acc + value, 0)
    const average = sum / validScores.length

    if (average < this.#alertThreshold) {
      return {
        status: 'ALERT',
        averageScore: average,
      }
    }

    return {
      status: 'OK',
      averageScore: average,
    }
  }
}
